### PR TITLE
Feature/add jenkinsfile generator

### PR DESCRIPTION
**Pull request by shouvSarker**
Added configMap generation function

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -124,6 +145,28 @@ To generate docker compose output for a project, run:
 
 This will generate a docker compose file at the specified path.
 
+Jenkins Pipeline File Template
```
Comment (dspasojevic) [inline README.md:172]: Is the command to generate the config maps documented?

File: src/main/scala/au/com/agiledigital/toolform/command/generate/kubernetes/jenconfig/GenerateJenkinsConfigFilesCommand.scala [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+  def runGenerateConfigFiles(name: String, namespace: String, instance: String, volume: String, repo: String): Either[NonEmptyList[ToolFormError], String] =
+    for {
+      configFileStatus   <- writeConfigFile(name, namespace, instance)
+      dispatchFileStatus <- writeDispatchFile(name, namespace, repo, volume)
+    } yield s"$configFileStatus\n$dispatchFileStatus"
+
+  def writeConfigFile(name: String, namespace: String, instance: String): Either[NonEmptyList[ToolFormError], String] = {
+    val configFile         = "configMaps/config_map.yml"
+    val configFileTemplate = "configMaps/config_map.yml.mustache"
+    val templateMapping    = Map("name" -> s"$name", "namespace" -> s"$namespace", "instance" -> s"$instance")
```
Comment (dspasojevic) [inline src/main/scala/au/com/agiledigital/toolform/command/generate/kubernetes/jenconfig/GenerateJenkinsConfigFilesCommand.scala]: Why is string interpolation used here?

File: src/main/scala/au/com/agiledigital/toolform/command/generate/kubernetes/jenconfig/GenerateJenkinsConfigFilesCommand.scala [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    val buildConfigFile = engine.layout(
+      configFileTemplate,
+      templateMapping
+    )
+
+    try {
+      writer.write(buildConfigFile)
+      Right(s"Wrote the config map into $configFile")
+    } finally {
+      writer.close()
```
Comment (dspasojevic) [inline src/main/scala/au/com/agiledigital/toolform/command/generate/kubernetes/jenconfig/GenerateJenkinsConfigFilesCommand.scala:82]: Does the rest of the code base model catch and return exceptions as Lefts?

File: src/main/scala/au/com/agiledigital/toolform/command/generate/kubernetes/jenconfig/GenerateJenkinsConfigFilesCommand.scala [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    }
+  }
+
+  def writeDispatchFile(name: String, namespace: String, repo: String, volume: String): Either[NonEmptyList[ToolFormError], String] = {
+    val dispatchFile         = "configMaps/dep_patch.yml"
+    val dispatchFileTemplate = "configMaps/dep_patch.yml.mustache"
+    val templateMapping      = Map("name" -> s"$name", "namespace" -> s"$namespace", "repo" -> s"$repo", "volume" -> s"$volume")
+
+    val writer = new BufferedWriter(new FileWriter(dispatchFile, false))
+    val buildDispatchFile = engine.layout(
```
Comment (dspasojevic) [inline src/main/scala/au/com/agiledigital/toolform/command/generate/kubernetes/jenconfig/GenerateJenkinsConfigFilesCommand.scala]: Can this fail and cause the `writer` to remain open?

File: configMaps/config_map.yml.mustache [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
...LtLyumRfDrPAXx7fQPuXoRGtKDQPXIGol6xkeZPvNhPXBnvopUQUBqzhSv1s2DPykvzaSRzoE5KpcJf+eRY1eJ81jdnJZKQBjXDzr0xMYETOlVBz1xBrvUTT2dPb4Fsitm2YY4eXZj+izqrd5ljnOQTG60lrLasZndKYdkJk/A5YqTAh67z6wMzxvUoOhStcywmhdS7u9bx3f2Ebo9q8Lz8rRRH/e/RqG+al9GuXT0uzNc5xAEWTAH8KnVFk6acFQjcVC6QqvkYMM5Dzwqu9EfwtMboNgQQDfUL84+vBfX1ZoUZSFTAQUmhqRVX5RCcOHB3ywxbAkgNn6duSzYbLmSmbSrlVe8We5DDyjMr4a3mTmBelw12KRTKSkF9mcaFLPvW4UplpEIbPWI79LRtE5xcFEsKQNRZ4PYqDS2/weyy0QbQinOtMH7Xtq11W6qk9uDJCyJeipU/WEP3BqbeTgFmffVZiR48vTw+KY5PqC/FNELN6ErABFfKPT4VxUBxr60L6ngMmZLlE6C8b+oA+pt40vimlcBZ+7Xp6mZ5RbJX8JC3y+BvfUoDhaOZ82JqfpVqE+BZAd3uRCvQZcdfagfUTQSYHlALix+cozV1X1FS7VhkMKef3Mtj5b0v3h3hY/45Z0x9ngkKHII3ZPVSjAbGLuYX2iVoCyR80MksN1cx+cYYCdOY3fH+9zleU9Pakc81B4aaequ0TVYX2+VDCH8Wu/aKvlnNi5h1M00LxfSnJmRhptyNTYWZByDorG6sksl995D5nygeV3aitoAe/HJou2k93J+gr8DdW9scm0MsEpYINxFg+U60TmVv6/g5pS3Le6+BPttKqjoPUBaiwlRrxU3pPQ0HbxVWqoOSD2KJh2nX6F48vHW3pw+cx4jTvfvztyzEw2j9F42tQxczdoNGLDnsOUc7m77acP1bJ+0LSHPUZmA/62FIP5Iryzqurpz67nuU88cy1xo4kkQGmDSUeq4AtqJVP8gxEFtluTHfzGXn6csHsmZpIz2dzEvlky+w9zUDMxtruQl001SriRpSEqoRdkDJxHgfUZKv6ohW/vyk41Pwj0oDaK8s+m2kTviWmQR/FvXPpGGp+pVTJmDJmrKVCFAWGG7n2NtwLTRoR2EKBFfIHzWaZSGWiK5/Gzs7TcaD89+i2w54yk6xrnq0+njofQYlUU=}</privateKey>
```
Comment (dspasojevic) [inline configMaps/config_map.yml.mustache]: This doesn't look very general...

**Review by dspasojevic — CHANGES_REQUESTED** [review]
Looking good. A few minor questions. The one about the specificity of the config map template is the most important.

File: README.md (hunk 1 above)
Comment (shouvSarker) [reply README.md:172]: Added documentation for the command to generate the config maps in readme

File: src/main/scala/au/com/agiledigital/toolform/command/generate/kubernetes/jenconfig/GenerateJenkinsConfigFilesCommand.scala (hunk 2 above)
Comment (shouvSarker) [reply src/main/scala/au/com/agiledigital/toolform/command/generate/kubernetes/jenconfig/GenerateJenkinsConfigFilesCommand.scala]: Removed the string interpolation. My bad, it was not necessary.

File: configMaps/config_map.yml.mustache (hunk 5 above)
Comment (shouvSarker) [reply configMaps/config_map.yml.mustache]: Replaced with a relatively generic version of the template in latest commit.

File: src/main/scala/au/com/agiledigital/toolform/command/generate/kubernetes/jenconfig/GenerateJenkinsConfigFilesCommand.scala (hunk 3 above)
Comment (shouvSarker) [reply src/main/scala/au/com/agiledigital/toolform/command/generate/kubernetes/jenconfig/GenerateJenkinsConfigFilesCommand.scala:82]: Yes, followed GenerateJenkinsFilesCommand.scala when writing this part of the code, and there exceptions are returned as lefts.

```
    if (!inputFile.exists()) {
      Left(NonEmptyList.of(ToolFormError(s"Input file [$inputFile] does not exist.")))
    } else if (!inputFile.isFile) {
      Left(NonEmptyList.of(ToolFormError(s"Input file [$inputFile] is not a valid file.")))
    } else if (!buildFile.getParentFile.exists()) {
      Left(NonEmptyList.of(ToolFormError(s"Output directory [${buildFile.getParentFile}] does not exist.")))
    } else {
```

File: configMaps/dep_patch.yml.mustache [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+  template:
+    spec:
+      serviceAccountName: {{namespace}}
+      initContainers:
+        - name: copy-default-config
+          volumeMounts:
+            - mountPath: /var/jenkins_jobs/jobs/{{repo}}
+              name: {{volume}}
+      volumes:
+      - name: csc-shouv-jenkins-job-test-jenkins-volume
```
Comment (dspasojevic) [inline configMaps/dep_patch.yml.mustache]: `shouv`?

File: configMaps/config_map.yml.mustache [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+              <scm>git</scm>
+            </com.cloudbees.jenkins.plugins.bitbucket.BitbucketRepoMetadataAction>
+            <com.cloudbees.jenkins.plugins.bitbucket.BitbucketDefaultBranch plugin="cloudbees-bitbucket-branch-source@2.6.0">
+              <repoOwner>CSC</repoOwner>
+              <repository>${REPO_NAME}</repository>
+              <defaultBranch>develop</defaultBranch>
+            </com.cloudbees.jenkins.plugins.bitbucket.BitbucketDefaultBranch>
+            <com.cloudbees.jenkins.plugins.bitbucket.BitbucketLink plugin="cloudbees-bitbucket-branch-source@2.6.0">
+              <iconClassName>icon-bitbucket-repo</iconClassName>
+              <url>https://stash.agiledigital.com.au/projects/CSC/repos/${REPO_NAME}</url>
```
Comment (dspasojevic) [inline configMaps/config_map.yml.mustache]: Please don't hardcode this.

File: configMaps/config_map.yml.mustache [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+    <jenkins.branch.MultiBranchProject_-State plugin="branch-api@2.5.4">
+      <sourceActions>
+        <entry>
+          <string>9c19f900-3f7b-4f13-818d-7e0088101389</string>
+          <list>
+            <com.cloudbees.jenkins.plugins.bitbucket.BitbucketRepoMetadataAction plugin="cloudbees-bitbucket-branch-source@2.6.0">
+              <scm>git</scm>
+            </com.cloudbees.jenkins.plugins.bitbucket.BitbucketRepoMetadataAction>
+            <com.cloudbees.jenkins.plugins.bitbucket.BitbucketDefaultBranch plugin="cloudbees-bitbucket-branch-source@2.6.0">
+              <repoOwner>CSC</repoOwner>
```
Comment (dspasojevic) [inline configMaps/config_map.yml.mustache]: `CSC`?

File: src/main/scala/au/com/agiledigital/toolform/command/generate/kubernetes/jenconfig/GenerateJenkinsConfigFilesCommand.scala (hunk 4 above)
Comment (shouvSarker) [reply src/main/scala/au/com/agiledigital/toolform/command/generate/kubernetes/jenconfig/GenerateJenkinsConfigFilesCommand.scala]: There is a try finally when writer is writing, and the bufferedwriter is closed in finally, so in both scenarios of the writer (failing or passing), writer will be closed

File: configMaps/dep_patch.yml.mustache (hunk 6 above)
Comment (shouvSarker) [reply configMaps/dep_patch.yml.mustache]: Restructured the entire config map template to get rid of similar issues

File: configMaps/config_map.yml.mustache (hunk 7 above)
Comment (shouvSarker) [reply configMaps/config_map.yml.mustache]: Removed hardcoding. Github/bitbucket specifics are now supplied by users in jenkins

File: configMaps/config_map.yml.mustache (hunk 8 above)
Comment (shouvSarker) [reply configMaps/config_map.yml.mustache]: Removed all project specific content

**Pull request closed without merging**